Funkcja regresji liniowej.

In [3]:
import pandas as pd
import numpy as np
#from bigfloat import *
import sympy as smp

data = pd.read_csv('genomes.csv')

x = data['size'].values
y1 = data['GC'].values

y = np.log(y1)

def mean(values):
    return sum(values) / float(len(values))

def variance(values, mean):
    return sum([(x-mean)**2 for x in values])

mean_X, mean_Y = mean(x), mean(y)
var_x, var_y = variance(x, mean_X), variance(y, mean_Y)

def covariance(x, mean_x, y, mean_y):
    covar = 0.0
    for i in range(len(x)):
        covar += (x[i] - mean_x)*(y[i] - mean_y)
    return covar

def coefficients(data,x,y):
    mean_x, mean_y = mean(x), mean(y)
    b1 = covariance(x, mean_x, y, mean_y) / variance(x, mean_x)
    b0 = mean_y - b1 * mean_x
    return [b0, b1]

z = coefficients(data,x,y)

def convert(x):
    c=x*(10**14)
    return c

a = convert(z[0])
b = convert(z[1])
print(f'b0:{a}, b1:{b}')

print('Zaokrąglony wynik b0:',round(a))
print('Zaokrąglony wynik b0:',round(b))

print('Zaokrąglony wyniku do dwóch miejsc po przecinku:',format(a,".2f"))
print('Zaokrąglony wyniku do dwóch miejsc po przecinku:',format(b,".2f"))


from decimal import *
print('Korzystanie z moduły decimal b0:',Decimal(a))
print('Korzystanie z moduły decimal b1:',Decimal(b))


from mpmath import *

mp.dps = 50
print('Z użyciem mpmath(mpf) b0:',mpf(a))
print('Z użyciem mpmath(mpc) b0:',mpc(a))
print('Z użyciem mpmath(mpf) b1:',mpf(b))
print('Z użyciem mpmath(mpc) b1:',mpc(b))

import gmpy2
from gmpy2 import mpfr
print('Z użyciem mpfr dla b0:',mpfr(a))
print('Z użyciem mpfr dla b1:',mpfr(b))

gmpy2.get_context().precision=333
print('Z użyciem mpfr ze zmianą precyzji dla b0:',mpfr(a))
print('Z użyciem mpfr ze zmianą precyzji dla b1:',mpfr(b))

print('Z użyciem String Formatting Operator,b0:',format(a, '.15f'))
print('Z użyciem String Formatting Operator,b1:',format(b, '.15f'))


x=smp.Float(a)
x1=smp.Float(b)
print('Z użyciem sympy(smp) b0: ',x)
print('Z użyciem sympy(smp) b1: ',x1)

print('Z użyciem biblioteki numpy(longdouble),b0:',np.longdouble(a))
print('Z użyciem biblioteki numpy(longdouble),b1:',np.longdouble(b))
print('Z użyciem biblioteki numpy(double),b0:',np.double(a))
print('Z użyciem biblioteki numpy(double),b1:',np.double(b))
print('Z użyciem biblioteki numpy(clongdouble),b0:',np.clongdouble(a))
print('Z użyciem biblioteki numpy(clongdouble),b1:',np.clongdouble(b))

print('Z użyciem String Formatting Operator dla b0: %e'%a)
print('Z użyciem String Formatting Operator dla b1: %e'%b)

b0:350431975154361.1, b1:9099230808282.246
Zaokrąglony wynik b0: 350431975154361.0
Zaokrąglony wynik b0: 9099230808282.0
Zaokrąglony wyniku do dwóch miejsc po przecinku: 350431975154361.12
Zaokrąglony wyniku do dwóch miejsc po przecinku: 9099230808282.25
Korzystanie z moduły decimal b0: 350431975154361.125
Korzystanie z moduły decimal b1: 9099230808282.24609375
Z użyciem mpmath(mpf) b0: 350431975154361.125
Z użyciem mpmath(mpc) b0: (350431975154361.125 + 0.0j)
Z użyciem mpmath(mpf) b1: 9099230808282.24609375
Z użyciem mpmath(mpc) b1: (9099230808282.24609375 + 0.0j)
Z użyciem mpfr dla b0: 350431975154361.125
Z użyciem mpfr dla b1: 9099230808282.24609375
Z użyciem mpfr ze zmianą precyzji dla b0: 350431975154361.125
Z użyciem mpfr ze zmianą precyzji dla b1: 9099230808282.24609375
Z użyciem String Formatting Operator,b0: 350431975154361.125000000000000
Z użyciem String Formatting Operator,b1: 9099230808282.246093750000000
Z użyciem sympy(smp) b0:  350431975154361.
Z użyciem sympy(smp) b1: 

Wartość estymatora największa jest dla funkcji mpf z biblioteki mpmath,z użyciem funkcji mpfr oraz decimal,wynosi  dla b0:350431975154361.125 oraz dla b1: 9099230808282.24609375

Wartość estymatora najmniejsza jest dla funkcji smp z biblioteki sympy,wynosi dla b0: 350431975154361. oraz dla b1: 9099230808282.25 

Różnice pomiędzy wartościami: polegają na zaokrąglaniu liczb,użyte funkcje są bardziej precyzyjne.


Kod z użyciem biblioteki numpy:

In [28]:
import numpy as np
import pandas as pd


data = pd.read_csv('genomes.csv')

X = data['size'].values
Y = data['GC'].values

mean_x = np.mean(X)
mean_y = np.mean(Y)

m = len(X)

numer = 0
denom = 0
for i in range(m):
    numer += (X[i] - mean_x) * (Y[i] - mean_y)
    denom += (X[i] - mean_x) ** 2
b1 = numer / denom
b0 = mean_y - (b1 * mean_x)
print(f'b0:{b0}, b1:{b1}')
'''
rmse = 0
for i in range(m):
    y_pred = b0 + b1 * X[i]
    rmse += (Y[i] - y_pred) ** 2
rmse = np.sqrt(rmse/m)
print('RMSE:',rmse)
'''

b0:32.68246972988583, b1:4.252880674352682


"\nrmse = 0\nfor i in range(m):\n    y_pred = b0 + b1 * X[i]\n    rmse += (Y[i] - y_pred) ** 2\nrmse = np.sqrt(rmse/m)\nprint('RMSE:',rmse)\n"